DC: Use regular ScummVM menu

tsowell opened the pull request:
This replicates the functionality of the Dreamcast game menu using regular ScummVM GUI widgets.  Dialog widgets replace the plugin loading progress messages, LauncherDialog replaces the game menu, and MassAddDialog scans the CD for games.

To support this, MassAddDialog now has a (default false) "autoClose" flag which indicates that it should close immediately upon completion instead of waiting for user interaction.

The UI experience is similar to the current game menu:

  * Upon booting, plugins are loaded from the CD drive root.  If there are no plugins in the root directory, a BrowserDialog prompts to choose a different plugin directory.  During plugin loading, Dialogs display the plugin filename and the amount of memory available.

  * Before the launcher is displayed, a MassAddDialog scans the CD for games and automatically closes and adds them to the game list.

  * The game list updates when the lid is opened (to disable the games on the CD).  When the lid closes, a new MassAddDialog scans the CD for games and updates the game list with any new games.

The main impetus for this is the SD card support I'm working on contributing - a lot of the benefits of the full ScummVM menu are unlocked with a writable config file.  There are still advantages for CD-only use, including:

  * Consistency with other platforms

  * The ability to change options, even if they aren't saved for the next session

  * It fixes bug #7704 (DC: Add game chooser when more than one game detected)

That being said, I'm new here, and I know this is a big change and might not be the direction you want to take the Dreamcast port.  Let me know if there's a better way to do any of this or if I should take a different approach.

digitall commented:
@tsowell : Thanks for your contribution. This looks very interesting and as you indicate, it fixes a few bugs and makes the DC port more consistent with other ports. The issue with the DC is always the limited RAM available ie. 16Mb, so I think previously it has been a limited GUI to avoid too much overhead, though that may not be a "real" issue.

Can I ask if you have played any games with this build? And if so, which ones and there was sufficient RAM to finish them! :)

I have triggered a request for review by @zeldin who is our original DC porter, but overall I think this looks like a good change... assuming everything still runs without OOM :)

zeldin commented:
There are a couple of issues to consider here.

The first is the memory situation.  The issue here is not so much that the default GUI is taking a lot of memory, but rather that this change removes the plugin directory chooser, which is required for a distribution on a single disc to be possible.  If you put all the plugins on a disc and scummvm loads all of them, you will run out of memory from just loading the plugins.  So you can't even choose a game, much less start it.  Thus I added the feature that you can partition the plugins into groups, and select a group when booting and only those plugins would be loaded.  Previously I solved this by authoring multiple CDs with the different plugin sets, but the situation was becoming unwieldy.

The second is the lack of persistent storage.  At least historically (I haven't checked in a while) the regular GUI worked by having a separate game scan function which saved the results for later use. something which does not work well on the DC (while you could save to a memory card, the same memory card might not be available on the next run), and also the concept of a "library" of available games does not map so well to a situation when you swap game discs in and out.

So while I haven't tested the patch yet. I feel that it is important to make sure that the following points are preserved:
* It is possible to author a single CD with all the plugins on it and use it to play games on other disks, with only a single disk swap
* You do not have to perform extra UI operations to add the game to your library every time scummvm is booted
* Swapping game disks works well, and does not confuse the user as to which games can currently be launched

zeldin commented:
Oh, one more thing:
* The custom save icons for games that support them (e.g. Sam & Max) should still work

digitall commented:
@zeldin: Thanks for those comments and background which should be useful to @tsowell. Hopefully you can get time to review / test this over the next few weeks, or maybe over Xmas break... I suspect I will end up working on bits then and avoiding awful TV :)

tsowell commented:
@digitall: I don't think this should have an affect on the memory available to games, as the launcher gets popped off the stack before a game is launched, but I'll do some testing to measure the actual impact.  I've only played through qfg1vga using the launcher (but not this build exactly).


@zeldin: Thanks for taking a look and for the feedback.  I tried to make this behave as close as possible to the current menu, and I believe most of these have been addressed.

> It is possible to author a single CD with all the plugins on it and use it to play games on other disks, with only a single disk swap

Yes, this works very similar to the current Dreamcast menu.  If no plugins are found in the CD's root directory, it displays a BrowserDialog prompting to select a plugin directory.  If the user chooses a directory, plugins are loaded from only that directory.  From there, they can load a new CD and play games from it.

> You do not have to perform extra UI operations to add the game to your library every time scummvm is booted

No user intervention is required.  When ScummVM boots, a MassAddDialog pops up automatically and scans for games.  When it has finished scanning, it automatically closes and adds all the games to the library.

> Swapping game disks works well, and does not confuse the user as to which games can currently be launched

I think swapping disks works well.  When the user opens the CD lid, any library games on the current disk are disabled and turn grey.  When the user closes the CD lid, the new CD is scanned, and the games are either re-enabled or added to the menu.

> The custom save icons for games that support them (e.g. Sam & Max) should still work

I had not tested the save icons, and they were not working.  I've added a new commit that finds and loads a game's icon before starting the game, and they work now.

By the way, thanks for all your Dreamcast work over the years.  I've been using your web site as a reference for about 15 years now, and it has been enormously helpful.

zeldin commented:
@tsowell: It sounds like you've done a solid job. :100:  I'll try to take a closer look and fire it up myself next weekend (this one was already fully booked unfortunately).

zeldin commented on backends/platform/dc/dclauncher.cpp:
```diff
@@ -0,0 +1,123 @@
+	if (!gameid.empty()) {
+		reportf("path %s\n", ConfMan.get("path").c_str());
+		Common::String path = ConfMan.get("path");
+		sprintf(icofn, "%s%s.ICO", path.c_str(), gameid.c_str());
```
Use snprintf here.
```suggestion
		snprintf(icofn, sizeof(icofn), "%s%s.ICO", path.c_str(), gameid.c_str());
```

zeldin commented on backends/platform/dc/dclauncher.cpp:
```diff
@@ -0,0 +1,123 @@
+	// Set the icon if a game has been selected
+	Common::String gameid = ConfMan.get("gameid");
+	if (!gameid.empty()) {
+		reportf("path %s\n", ConfMan.get("path").c_str());
```
Use of reportf should be guarded with ``#ifndef NOSERIAL``

tsowell commented:
Thanks for the feedback, @zeldin.

I think there are a couple of options for setting the plugin dir selector starting path.  The BrowserDialog widget uses the browser_lastpath configuration key for the starting path, so one way would be to include in the Nero image's filesystem a scummvm.ini file with browser_lastpath set to "/PLUGINS".  Or we could modify BrowserDialog so that the caller can optionally provide a starting path - I'll start looking into this approach.

I have run into "Not enough space available!" too, and the solution for me was to make sure that ScummVM was building with zlib support.  Apparently many games' saves are too large to fit on a VMS without compression.

zeldin commented:
Well, the thing is that I have an earlier save from Sam & Max which is "only" 56 blocks, so if >200 blocks (impossible to satisfy) is needed now that's a regression.

tsowell commented:
When I experienced this problem, it was because I had compiled ScummVM without zlib support.  Scumm saves tend to be about 100K in size, so they don't fit on a VMS with compression.  When ScummVM is compiled with zlib support, it compresses saves down to the 10-20K range which does fit on a VMS.  I wonder if your 56-block save was created with a build that was compiled with zlib support, and you're currently testing with a build compiled without zlib support.

digitall commented:
@zeldin : You could just restore the hardcoded enablers, but if we do, can you add a comment in the configure script as to why that is needed. I think the "correct" solution would be to patch cc_check to respect the $INCLUDES ... but that could have side effects for other ports, but could try it and commit to see if the buildbot works fine at least... though not sure if that needs to be a PR.

digitall commented:
@tsowell : This PR now has some conflicts due to commit 79a4e3f8131fa8e3f2b3041e2a440236093cb76d which touched the gui/options.cpp defines to remove the GP32 backend. I could probably resolve correctly, but if you could take a look and resolve that would probably be better :)

zeldin commented:
@digitall: Yeah, my local fix was to add $INCLUDES (and $DEFINES) to ``cc_check_no_clean``, because that's how ``Makefile.common`` is going to compile stuff later so it is the environment it _should_ test in.  :smile:
This fix makes detection work correctly for me and I was planning on pushing it, but if you want to test it on some other ports first you are very welcome.  :cake:

zeldin commented:
@digitall : zeldin@baa1f2d97f150c92e4bb39ad28ae145079da1f2a

digitall commented:
@zeldin : That commit looks good to me. To be honest, I tend toward "publish and be damned" so I would just push it as probably the only way any minor port issues will get thrown up... and if they do they probably should be fixed anyway.

tsowell commented:
@zeldin, you're right - numbers are added to the end of domain names to ensure their uniqueness, and if a base gameid is already longer than 6 characters, it will exceed the 8-character limit (https://bugs.scummvm.org/ticket/10410).  I have some ideas for how that could be fixed, but it's not going to be easy and is outside the scope of this PR.

For now I will try to duplicate the existing behavior to keep things consistent.

tsowell commented:
@digitall, sounds good.  I'll rebase and retest the PR once I've addressed @zeldin's feedback.

digitall commented:
@tsowell : Sounds like a plan. Thanks.

zeldin commented:
@tsowell : Yeah, even if some trick would be employed to handle >12 character filenames in general, we still need the savegame filenames to be consistent and not depend on which discs were swapped in/out before launch.  Thanks.

zeldin commented:
@digitall : Looks like I broke the mingw-w32 build on buildbot.  It doesn't detect any libs anymore, but AFAICT ``config.log`` is not available in the web ui, so I don't see why exactly...
mingw does not use $INCLUDES, but it does use $DEFINES.

digitall commented:
@zeldin : Will take a look ...

lotharsm commented:
@zeldin @digitall:

Latest config.log from the mingw-w32 build slave: https://gist.github.com/lotharsm/62da0c5995c0034c0481cd5c21efd2cd

digitall commented:
Odd the breakage appears to be in building the test runner, hence why it only affects the native ports (since tests can only run on native... since only native binaries are runnable):
https://buildbot.scummvm.org/builders/master-mingw-w32/builds/460/steps/test/logs/stdio

This might clear after a full clean / ccache flush when the nightly runs... so will see if this resolves.

digitall commented:
Even if it resolves, that doesn't mean it is detecting all libraries correctly, just that the build is not broken due to incremental build inconsistency.

zeldin commented:
Hm, actually, on closer inspection, I think the problem is that ``-Dmain=SDL_main`` enters ``$INCLUDES`` from ``sdl-config --cflags``, while the ``-lSDL2main`` needed for that to work goes into ``$LIBS``, which is _not_ added by ``cc_check_no_clean``.  Then the fix should be pretty straightforward (add ``$LIBS``)...

zeldin commented:
Nope, it still looks cranky.  Would be nice if someone with access to a MinGW system could take a gander...

digitall commented:
@zeldin Since the zlib issues seem to have been fixed... at least for now, happy for this to be fixed up for gui/options.cpp conflict and then merged?

zeldin commented:
@digitall : The issue with "-1" in the savegame names needs sorting out too, then it's good to go.

tsowell commented:
@digitall : Yep, thanks.  Last week was a little hectic, but I should be able to spend some time on it this week.  I'll let you know when it's ready for review.

digitall commented:
@tsowell : No problem! We all have day jobs :) Thanks again for looking at this.

tsowell commented:
@digitall and @zeldin : I think this is ready:

  * I addressed the code quality concerns about reportf and snprintf
  * The plugin directory selector starts in /plugins (if present)
  * Save filenames use gameid instead of the domain name (to reproduce existing behavior and to resolve the "-1" issue)
  * I rebased to resolve the merge conflict and retest

Let me know if you have any other concerns.  Otherwise, thank you both for your help!

digitall commented:
@tsowell : Looking damn good. It should be fine to merge as-is, but something that would be good to avoid is using platform specific defines within the engine(s) as these are meant to be abstracted by OSystem. The various changes in there with ifndef __DC__ seem to be related to switching ConfMan.getActiveDomain to instead return gameid... Could this change be done without ConfMan in a single location and not within engines at least? i.e. change ConfMan.getActiveDomain return value for _DC_ in a single location rather than in the engine base class and several other places across engines ... Just that if we don't this is likely to get broken in future by another engine lacking this special define... and it is just not very backend vs. engine separation clean.

zeldin commented:
@tsowell: Thanks!  I do agree with digitall that it would be nicer if you did not have to touch engine code.

I don't know if this is really important, but did you consider the case of a disc authored with a custom scummvm.ini on it?  I _think_ this used to work, provided you used the gameid as target, so it would be nice if it continued to do so.  Not sure if anyone is actually using it...

ccawley2011 commented on backends/platform/dc/dclauncher.cpp:
```diff
@@ -0,0 +1,127 @@
+
+static bool isIcon(const Common::FSNode &entry)
+{
+	int l = entry.getDisplayName().size();
```
Previously, commit b224b59 changed this to use hasSuffixIgnoreCase(). Is there any reason why this was changed back?

zeldin replied:
I don't know if there was a particular reason, but it doesn't really matter if the comparison is done ignoring case or not; filenames on ISO-9660 are always uppercase...

tsowell commented:
Hi all, sorry to leave this hanging.  While it should be easy enough to reproduce the existing save game naming behavior, it would be pretty messy.  I think this should be shelved until the existing behavior is corrected.

bluegr commented:
The best way to solve the file naming issues here would be to use a separate SaveFileManager for the DC backend, and override the openForSaving() method so that it creates the save file names as expected by the file system

zeldin commented:
@bluegr The DC backend already has its own SaveFileManager.  The issue here is that the filename sent to openForLoading et al is different after this change (and also dependent on which order you insert discs) than before.

bluegr commented:
@zeldin: Right. The reason I missed it is that the DC's savefile manager does not inherit from the default one (inside backends/saves/default), and it resides in the dc backend directory (i.e. inside backends/platform/dc). Also, it is named differently (methods inside vmsave.cpp), and the methods are different as well (e.g. trySave(), tryLoad() etc). This should be adapted in the future.

It's not really efficient to adapt all the engines that are calling runModalWithPluginAndTarget() for DC. The savegame filename limitation for DC should be either in runModalWithPluginAndTarget(), or inside the vmsave methods, and not in engine code

zeldin commented:
@bluegr Again, the real issue here isn't the savefile filename limitation.  I know how to deal with that.  The issue is that if we accept this PR then people will not be able to load their savegames because the filenames will change, and keep changing, in a manner that violates the principle of least surprise.

For example, if they have previously played a game with gameid 'foo', then they will have savegames with namesd like 'foo.s01'.  But after the change ScummVM might instead look for 'foo-1.s01' or 'foo-2.s01' (and switch inconsistently between them on different runs, depending on how you start the game).  These filenames can be handled just fine, but are not the filenames used previously so the user won't find their saves.

That is what the changes to the engines are trying to fix, and while I agree that it is not the correct way to fix it, it don't quite follow what the vmsave methods could do about it.

zeldin commented:
@bluegr You are consistently missing the point.  The main issue is not with filename limitations, but with the fact that a suffix is added to differentiate between different "installations" of the game.  This works fine if the mapping between suffix and "installation" is kept constant in nonvolatile memory (which is what happens on other ports), but on the Dreamcast there is no location where such a mapping can be reliably stored.  Before this PR only one installation was added to the (volatile) config database (after the custom selector had done its thing), so a suffix was never added.  Thus the same installation would always get the same suffix (i.e. none).  But by using the regular selector, installations are added to the (volatile) config database as detection happens, which means that not only can the games get suffixes, they can get inconsistent suffixes depending on the order in which installations were added to the config database.

I think the issue would be mostly resolved if the database was cleared (and reloaded from scummvm.ini on the CD if one exists) on disc swap.  Then the only change would be that if you have two installations with the same gameid on the same CD, they would no longer share a savegame pool.  It could still make old savegames unusable with things like the dual language Loom CD though, in case the "wrong" installation gets the suffix.

Just replacing the filename with `ConfMan.get("gameid") ` is obviously not going to work because a game can have more than one savegame so the gameid/targetid is just one part of the filename.  Trying to parse the given filename to isolate the different parts and replace them selectively could ostensibly work, but seems somewhat fragile...

sev- commented:
It seems that this nice PR got kind of stuck.

One of the ways of addressing saves would be to transparently truncate it, or even use first 2 characters, followed by CRC32 of the filename (without extension), thus, `samnmax-1.s01` becomes `sa22D74833.s01` and `samnmax.s01` stays as `sa5B280160.s01`. Then, when you need to scan a directory for saves when querying them for enlisting in GUI, you will have a consistent way of replacing filename with proper CRC.

The only obvious disadvantage is that you losing the possibility to take the save from DC and move it to your desktop as is, but something tells me that this process is so special and requires additional hardware, so it is not practical.

zeldin commented:
@sev- A much more severe disadvantage of your proposal is that you lose the possibility of continuing a game that you saved on the DC with an earlier version of ScummVM...  It is this possibility I've been saying all along that I want to keep...

sev- commented:
@zeldin, I do not understand what you're talking about:

```
 openSaveFile(char *s) {
    if (!file.open(s)) {
        Common::String newname = genFileName(s);
        if (!file.open(newname))
            return NotFound;
    }
    ...
}
```

And saving is always with generated file names.

zeldin commented:
Ok, sure, you can rename old saves to the new scheme, but then you instead lose the possibility of going back to the old scummvm version.  Not to mention that the "rename" function would have to be implemented in the first place...  :wink:

zeldin commented:
And anyway it's not solving the actual problem, which is that `s` will have the wrong contents ("samnmax-1.s01" instead of "samnmax.s01"), so `file.open(s)` would fail.

sev- commented:
@zeldin, I do not understand this last comment. I mean, no previous version could ever create `samnmax-1.s01` file, right? So, we will not have older saves, only newer.

Well, you could also use this scheme only in case your base filename exceeds 8. Thus, you will be able to have the previous saves untouched, and new ones, with longer names, created in a compatible way.

zeldin commented:
@sev- Since you seem to have only skimmed the previous discussion, I'll summarize it for you.  There are two related but different issues at play here.

#1: This PR changes the way the GameDomain is set for games.  With the old selector, the GameDomain would always be the gameId.  So even if you had three variants of Sam&Max, two on one CD and on on another CD, the GameDomain would always be `samnmax` regardless of which one of them you launched.  So savegames would be called `samnmax.s01` and so on.  While having overlapping savegame names for three instances of the same game might not be ideal, you can work around it by using different memory cards for the three.

With this PR, the GameDomain for these three instances would be `samnmax`, `samnmax-1`, and `samnmax-2`, and the savegames for two of them would suddenly change to `samnmax-1.s01` and `samnmax-2.s02`, making the old saves inaccessible.  This is a regression, and applying some kind of hash to reduce the length of `samnmax-01.s01` will not fix that, you need to look for `samnmax.s01` because that is the filename the old save will have.

Also note that the assigment of which instances get the `-1` and `-2` suffixes is volatile; if you insert the CD with two instances first and then the CD with one instance, then the latter will be called `samnmax-2`.  But if you power cycle the system and then insert the second CD first, the instance on that one will be called `samnmax`.  So newly created savefiles are also hit by this issue.

#2: If a game with a GameDomain longer than 8 characters is launched, then the savegame filenames will exceed the limit of 12 characters, and saving will not work.  This is not a regression; the old launcher did not perform a length on the gameId, so if it was something like `discworld`, that would break.  This would be nice to fix, but is not directly related to this PR.  Issue #1 merely exposed this issue by creating a longer GameDomain than 8 characters for a game where this used to not be the case.

After some thinking I'm beginning to lean towards a KISS solution here - simply remove the regexp `-[0-9]+` from the filenames in the SaveFileManager.  It's a bit hackish, but as long as no gameId contains a dash followed by a number (such as `monkey-2`), it should work.  If the filename is still to long after this stripping, then some hashing transformation could optionally be applied to address the second issue.

This assumes that engines will be ok with finding filenames such as `samnmax.s01` and `discw~a7.s01` in the result when they call `listSavefiles("samnmax-01.s*")` and `listSavefiles("discworld.s*")`.

sev- commented:
I see. Although I read the whole thread, somehow I missed it. Thanks for the sum up.

In this case I would add a platform-dependent code to `EngineManager::createTargetForGame()` in plugins.cpp:577, something like If it is DS, and a target exists, then remove the previous one.

As a result, if two variants of a game will sit on same CD, only the last one will be accessible, but I guess it is how it was working before.

zeldin commented:
Actually, having two variants on the same CD _did_ work before.  The trick is that no Target/GameDomain is created until _after_ the user picks the game to launch (and then only for that specific instance).  Before that only `EngineMan.detectGames()` is used, so there is no contention between multiple instances.

henke37 commented:
Would there be a point in waiting with this PR until after the GSOC project to redo the metaengine parts of the plugin system?

zeldin commented:
@henke37 Possibly, if the plan is also to refactor the savegame handling.  Right now there is two problems with solving this in the MetaEngine:
1) Not all engines use MetaEngine (by way of GUI::SaveLoadChooser) for savegame handling
2) The engine can pass a non-NULL string pointer to override the basename chosen by MetaEngine

zeldin commented:
@tsowell I have some time to work on this now, but it looks like the changes do not apply cleanly to `master` anymore.  Do you think you could take a look at fixing the conflicts, and I'll try to do something about the savefile filenames in the backend.

zeldin commented:
I now have a workaround for `backends/platform/dc/vmsave.cpp` which makes old saves work without commit 1d159cd1.  It's a little bit hackish, but it works (although it will break if you have more than 99 savegames in myst, because there is no way for it to know that `myst-100.mys` is not a save for the 100:th installation of myst, which should have been `myst.mys` by the old rules)...

tsowell commented:
> @tsowell I have some time to work on this now, but it looks like the changes do not apply cleanly to `master` anymore. Do you think you could take a look at fixing the conflicts, and I'll try to do something about the savefile filenames in the backend.

Hi @zeldin, sorry, but I don't know when or if I'll have time to spend on this.  I don't even have my notes for creating a build environment for this backend anymore, so I'd be starting from the very beginning.

sev- commented:
Do we have any update for this PR?

zeldin commented:
@sev- I have not heard anything more from tsowell.

tsowell commented:
Sorry if you've been waiting on me!  I'm afraid I don't have time to work on this.  Feel free to close it if no one else can pick it up.

sev- commented:
It is a pity. I'm closing this, but please do not purge your branch. Perhaps someone will pick it up in the future.
